upgrade deps - coverage, caching and source-mapping fixes, ...

huafu opened the pull request:
What does this PR:

- [x] ⬆️ upgrade `babel-core` => `@babel/core` and related modifications

    This will hopefully resolves some issues related to incompatible versions between babel and jest

- [x] ➖ `babel-plugin-istanbul` and `babel-plugin-jest`: dependencies => dev deps

    Those are included automatically when adding `babel-jest` and we should leverage it to choose the correct versions. Plus it allows users to use their desired versions if necessary

- [x] 🔨 uses `jest` core types
- [x] 🐛 fixes some coverage issues
- [x] 🐛 fixes line mapping issues #606 
- [x] ⬆️ upgrades `typescript` and `tslint` to latest versions (dev deps)
- [x] :bug: fixes caching issues (#524 and maybe others)

     I first thought this was related to jest itself, but then saw how the cache key was created and read some issues in this repo. Now the content of package.json and babel config if any are included in the hash builder input.

- [x] 🐛 fixes an issue when transforming html files that would contain `${...}` or backtilts (`)

__Possibly__ related issues: #601, #614

Those are breaking changes! It could still work (I've not tested it), but I expect that using this modules with babel < 7 might be broken now.

---

Some minor stuff to note:

- removed js entry points in root, all are ts in `src/`, entry point is `index` there
- refactored some files
- dropped many types to use the ones provided by jest package
- split tsconfig.json (one for editing and tests, another for build)
- cleaned-up a tiny bit some package scripts

huafu commented:
@kulshekhar can I remove `transpileIfTypescript` and its containing file? it's not used anywhere, nether exported. One test file is testing against it also.

kulshekhar commented:
@huafu wow, this is turning into a big PR :)

I'm at work right now and will likely only get time to review this on the weekend. In general, it's fine to remove any part that isn't being used and the associated tests.

Let me know if you have more questions. I'll try to answer the ones I can right away!

huafu commented:
Yeah I spent my day on it because I spent yesterday and the day before trying to configure ts-jest for it to work with latest babel... There are many issues (even not linked) that this PR addresses (partially or not), I've updated tests, wrote some new expectation, ... to match what I've changed/added. I think more tests needs to be added related to line numbers matching and caching but globally existing tests remained (almost) untouched.

Is there a Slack or somewhere we could chat about it when you'll be back? Because I need to go back working on my own project now (with `yarn link ts-jest` for now) and exchanging here will take ages. GMT+2 here (not at home but in France).

kulshekhar commented:
I'm on discord (kul#1309) and have joined the Reactiflux server

ahnpnl commented:
@huafu man you rock the whole code base 🎉

huafu commented:
Well, I'll leave it as a WIP for now and come back to it after the weekend. There are still issues with line mapping that I want to fix before this gets in (if approved).

huafu commented:
note: 76e27c1 fixes #606 and other source mapping issues

GeeWee commented:
Damn this thing is *massive*

huafu commented:
OK, I think this is ready for review.

I've tested with 10+ minimal repo given in some open issues, noted [here](https://github.com/kulshekhar/ts-jest/pull/626#issuecomment-407349590) the ones this PR **does** fix (added notes in the others the one fixed - probably with previous PR).

GeeWee commented:
How far did you and @kulshekhar get? I'd like to get this in, but I'm also of two minds whether we should split it into smaller prs to make it easier to review.

huafu commented:
@GeeWee the thing is, it's hard to split since it's been refactored so that it's a little bit more readable (I hope at least) while upgrading packages to fix issues I've found and improving typings.

Most of the changes **IN THE TESTS** are:
- imports (like `babel-preset-env` => `@babel/preset-env`, `babel-core` => `@babel/core`).
- wrong tsconfig `include` in a passing test which should actually fail
- snapshots added after replacing `toMatch(/some regexp/)` to `toMatchSnapshot()` (I double checked each snapshot btw, new or not)
- snapshot update since added tests in suite
- expected error message strings changed in new babel
- added tests
- added expectations
- changed arguments as function signature changed
- changed the `transfom` in jest config since entrypoint is not `/preprocessor.js` anymore but `/dist/index.js`

So "mostly" tests are kept as is and passing.

---

Then while writing [this](https://github.com/kulshekhar/ts-jest/pull/626#issuecomment-407349590) I've tested each repo I could find in the issues (15 now, I even tested 2 or 3 that I knew my branch were not addressing, mostly react native related ones).

Before any of them, in my ts-jest branch:
```sh
yarn build
yarn link
```

```sh
git clone https://ulr-of-repo
cd repo
yarn
jest --clearCache && yarn test # or jest if no script configured
# ensured it failed

yarn add --dev ts-jest jest # ensure to have latests version
jest --clearCache && yarn test # or jest if no script configured
# ensured it failed again with latest version

yarn link ts-jest
jest --clearCache && yarn test # or jest if no script configured
# ensure it passes
```

...adding tests when relevant

I'd be glad to make a call if needed (hmmm tho IDK where, skype is kinda odd now), so that I can go with you step by step on the diff and answer any of your questions.

GeeWee commented on src/preprocess.ts:
```diff
@@ -50,23 +44,22 @@ export function process(
 
   const transpileOutput = transpileTypescript(filePath, src, compilerOptions);
 
+  // FIXME: this should be done in the typescript source, else it's invalidating source maps
```
Good catch. Is there already an issue on this?

huafu replied:
I don't think so! at least did not see one. Tho IDK if comments can break a sourcemap, it needs some investigation, I just put that comment here as I couldn't research while coding, but then forgot it haha!

GeeWee replied:
Fair enough. Let's leave as-is for now. I don't think it'll break line-numbers as there's no line-shifts in the replaced code.

GeeWee commented on src/preprocess.ts:
```diff
@@ -28,8 +22,8 @@ export function process(
-  if (isHtmlFile && jestConfig.globals.__TRANSFORM_HTML__) {
-    src = 'module.exports=`' + src + '`;';
+  if (isHtmlFile && (jestConfig.globals as any).__TRANSFORM_HTML__) {
+    src = 'module.exports=' + JSON.stringify(src) + ';';
```
why are we calling json.stringify on something that's already a string?

huafu replied:
So that the generated js source export a string ;-) 
If you do not stringify it, it'll be like:

```js
module.exports = <div>
  this is my html
</div>;
```

but it's NOT any JSX in there just bare html

GeeWee commented on tests/__helpers__/jest-config.ts:
```diff
@@ -0,0 +1,210 @@
+
+const { resolve } = require.requireActual('path');
+
+function createJestConfig(
```
I think I see what you're trying to do with this, but after looking at the examples where this has actually been used, and how difficult this code is to fathom (took me a while) I don't think this level of abstraction is worth the cost of comprehending it. It either needs to go, or be implemented in a much cleaner way. I'd suggest we kill it for now, and then try to work it into another PR

huafu replied:
OOooooohhh yeah, that is where I forgot a todo! I wanted to cleanup that crap, you can say it, it's crap! I'll see what I can do to remove it but keep the tests of course

GeeWee commented on tests/__tests__/disable-sourcemap-support.spec.ts:
```diff
@@ -8,7 +8,8 @@ describe('Typescript errors', () => {
     const stderr = result.stderr;
 
     // The actual error is on line 18 - the line# being wrong here means sourcemaps are disabled.
```
this is much better.

GeeWee commented on tests/__tests__/get-cache-key.spec.ts:
```diff
@@ -1,47 +1,34 @@
-    expect(newHash).not.toBe(originalHash);
-  });
-
-  it('should change hash when tsconfig changes', () => {
```
why was this removed?

huafu replied:
Hmmm... I think I forgot something here yeah. Will add a test for when tsconfig changes.

GeeWee commented on tests/__tests__/html-transform.spec.ts:
```diff
@@ -1,19 +1,32 @@
+    expect(transformed).toMatchSnapshot('source');
+    expect(untransformed).toMatchSnapshot('untransformed');
+    // requiring the transformed version should return the same string as the untransformed version
+    expect(exported).toBe(untransformed);
```
I'm not very confident in this area of the codebase, perhaps @kulshekhar could weigh in here. It seems good to me but I have no clue.

huafu replied:
The `__TRANSFORM_HTML__` set to `true` is to make an html file become a module returning the html code as a string. Before it wasn't using `JSON.stringify` to make the module source. I've just fixed that issue (actually I think the fix has been merged in another pr before). Here the test has been improved to ensure that the source of the module is correct, the result of `reuqire()` it is correct, and that it matches with the `readFile()` content.

GeeWee commented on tests/__tests__/tsconfig-default.spec.ts:
```diff
@@ -13,78 +15,55 @@ describe('get default ts config', () => {
+      );
 
+      // snapshot would be enough here, but that adds a security in case we do not see it in a PR
       expect(config.module).toBe(ts.ModuleKind.ES2015);
```
CI won't pass if there's a snapshot error.

huafu replied:
Yup, but this is because @kulshekhar was concerned about us not seeing a snapshot change in a PR, like if we have done a `jest ... -u` and did not check the snapshot diff. Here it's a little test ahead that would fail anyway, even with the `-u`. It's a security in case we hadn't seen the diff of the related snapshot.

GeeWee replied:
Ah okay. I don't think I share his fears, but I don't see any harm in leaving it in.

GeeWee commented on package.json:
```diff
@@ -1,18 +1,19 @@
     "pretest": "npm run tslint && npm run clean-build",
     "test": "node scripts/tests.js",
-    "tslint": "tslint src/*.ts",
+    "test:unit": "jest --testRegex '' --testMatch '<rootDir>/src/**/*.spec.ts'",
```
I don't think this will correctly update results if the ts-jest sources are updated, only if the test sources are. What's it for exactly?

huafu replied:
You cannot use `testRegex` alongside `testMatch`, so I had to set testRegex to nothing else I'd have jest bailing. Anyway, I'll remove this `test:unit`. At first I wanted to move all unit tests next to the sources in `src` folder to keep them well separated from all other "real-world" tests, but at the end it was already a big pr so I kept it like that :-D

huafu commented on package.json:
```diff
@@ -1,18 +1,19 @@
-    "clean": "rimraf dist/**/* && rimraf tests/simple/coverage && rimraf tests/simple-async/coverage && rimraf tests/**/*/debug.txt && rimraf tests/**/node_modules",
+    "build": "tsc -p tsconfig.build.json",
+    "build:watch": "tsc -p tsconfig.build.json -w",
+    "test:nolint": "npm run clean-build && jest --clearCache && node scripts/tests.js",
```
Yup, at least until facebook/jest#6610 got fixed

GeeWee replied:
Fair point.

huafu commented on package.json:
```diff
@@ -59,48 +60,56 @@
   "peerDependencies": {
-    "babel-core": "^6.26.3 || ^7.0.0-0",
-    "jest": "^23.0.0 || ^24.0.0",
+    "babel-core": "^7.0.0-0",
```
It's in beta, not alpha. And they are not forced. I've tested some minimal repo not using 7 by just `yarn link ts-jest`, it was working.

Anyway, the big issue is the 😡 move of babel from `babel-xxx` to `@babel/xxx`, big mess in many repo depending on it. The `babel-core ^7.0.0` ensure that if the user has some packages depending on `babel-core` and some others on `@babel/core` it'll use the `@babel/core` one.

GeeWee replied:
Isn't this going to end up with a lot of unmet peerDependency warnings though? I know React Native's switched over to babel7 already, but I'm very hesitant about forcing it here.

huafu replied:
@GeeWee I'm not an expert in npm and deps. If you know a better way to handle this, help more than welcome!

The facts:
- if **anything** has `babel-core` or `babel-preset-env` (old versions) as a dep, even tho you'd try to `yarn add @babel/core` it will bail until you `yarn add babel-jest` (latest).
- `babel` cannot be a dependency, as in the special case where users will set `skipBabel: true`, but mostly so that users can install their own version, same as typescript, jest, ... or should we add them as deps, but locking their versions?

GeeWee replied:
No this is a hard issue. It depends whether @kulshekhar is okay breaking backwards compatibillity  (because it *definitely* will, for a lot of people)

huafu commented on package.json:
```diff
@@ -59,48 +60,56 @@
     "cross-spawn": "latest",
     "cross-spawn-with-kill": "latest",
     "doctoc": "latest",
+    "flat": "^4.1.0",
```
oh yeah, good catch, I forgot to remove it!

huafu commented on package.json:
```diff
@@ -59,48 +60,56 @@
     "@types/yargs": "^11.0.0",
-    "babel-core": "^6.26.3",
-    "babel-preset-env": "^1.7.0",
+    "babel-plugin-istanbul": "^4.1.6",
```
Yup, another thing we have to do is improve the way we run the tests. Another TODO.next ;-)

Anyway, I'll remove babel-plugin-istanbul, you're right. I think then `babel-jest` must be included, and both istanbul and preset env removed, right?

huafu commented on scripts/tests.js:
```diff
@@ -16,7 +16,7 @@ function getDirectories(rootDir) {
   const testsRoot = 'tests';
   const testCaseFolders = getDirectories(testsRoot).filter(function(testDir) {
-    return !(testDir.startsWith('__') && testDir.endsWith('__'));
+    return !/^(?:utils|__.+__)$/.test(testDir);
```
No, it's to NOT handle the `utils` folder as a test case folder, so that node_modules isn't copied in that one ;-)

GeeWee replied:
Right, there's a !.

huafu replied:
Sure. I could have done, just it was there and I saw some node_modules in there... I could have leave the test, doesn't hurt if there are git ignore folder created in there... but yeah, moving that where it should go is the best solution you're right.

huafu commented on scripts/tests.js:
```diff
@@ -51,6 +41,9 @@ function createIntegrationMock() {
 
 createIntegrationMock();
 
+// HACK: allow us to change the `startDir()` during tests
```
Aew, I actually forgot to remove this. In fact, I have to see this with @kulshekhar as this was to keep using `starDir()` but with a better detection on whether we are running our tests or not (the purpose of this special env var). Tho I also commented out the `startDir()`. I must chat more about this with @kulshekhar. The thing is that I reviewed totally the way `tsconfig` file is resolved, so there is no need to startDir() anymore. I also kinda isolated the tests so that they can't pickup the default `tsconfig.json` in the root of the project outside of the tests by renaming the `/tsconfig.sjon`.

For the `tsconfig` resolution (`jestConfig` is full as all missing props replaced by defaults/correct ones by jest upstream):

- jest config **has** `tsConfigFile`
    - `tsConfigFile` is asolute ➡️ use it directly
    - `tsConfigFile` has `<rootDir>` ➡️ replace `rootDir` and use the result
    - `tsConfigFile` is relative, use `path.resolve(jestConfig.rootDir, tsConfigFile)` _I wanted to use jestCOnfig.cwd but in multi-projects jest config it resolves to the path of the root jestConfig and not of each sub-project jest config, so `.` will refer to rootDir and not the path of jest config as it would break in multi-proects configs_
- jest config **does not have** `tsConfigFile`
    - starting from `jestConfig.rootDir`, try to find a `tsconfig.json`, is each parent dir. (talked with @kulshekhar about this: now it'll throw an error if it can't find a tsconfig file)

GeeWee replied:
Alright, so we're throwing this out?

huafu replied:
FYI the `startDir()` was returning `.` if not running tests, else `../../` or something like that

huafu commented:
@GeeWee a big thanks for your time reviewing this!

> I haven't read closely in the utils files. The diff is pretty large. I assume you've just moved the functions out into seperate files? Have you done anything else with them?

well yeah I basically split utils into individual purpose files. But I also refactored/fixed the way tsconfig file is resolved and the way the cache key is generated.

> I think we should consider using inline snapshots instead, as long as the snapshots aren't too huge - it's going to make reviewing changes easier.

agreed, but how is it working with TS? Already saw people reporting issues with js tests... maybe we should wait a bit for that, and more than anything, see how it handles ts files hehe.

> As you can tell, not a huge fan of the jest-cfg file. I think we should ditch that, and potentially create another PR with it.

what do you mean with the `jest-cfg` file?

GeeWee commented:
> @GeeWee a big thanks for your time reviewing this!
Thanks for submitting it!

> well yeah I basically split utils into individual purpose files. But I also refactored/fixed the way tsconfig file is resolved and the way the cache key is generated.
Can you mention exactly what you've changed? It's a rather large diff to go diving through if you've just changed a few places (incidentally, this is also why I'd prefer smaller PRs )

> agreed, but how is it working with TS? Already saw people reporting issues with js tests... maybe we should wait a bit for that, and more than anything, see how it handles ts files hehe.

Oh, I hadn't considered that all. Fair point.

> what do you mean with the jest-cfg file?
Sorry, the `jest-config.ts` in the testing suite.

huafu commented on src/postprocess.ts:
```diff
@@ -52,70 +41,96 @@ function createBabelTransformer(
   return createBabelTransformer(babelOptions);
 };
+
+function toArray<T>(iter?: Iterable<T> | null): T[] {
```
You meant `castArray`, yup I wasn't aware about castArray, will do now

huafu replied:
hmmm nope. `castArray` creates `[null]` and `[undefined]` if you give them an argument of that value.

ahnpnl replied:
can we use `Array.from` ?

ahnpnl replied:
I think he meant using `toArray` not `castArray`

huafu replied:
Well, any of them does what the `toArray` do here. From lodash docs:

```js
_.toArray({ 'a': 1, 'b': 2 });
// => [1, 2]
 
_.toArray('abc');
// => ['a', 'b', 'c']
 
_.toArray(1);
// => []
 
_.toArray(null);
// => []
```

```js
_.castArray(1);
// => [1]
 
_.castArray({ 'a': 1 });
// => [{ 'a': 1 }]
 
_.castArray('abc');
// => ['abc']
 
_.castArray(null);
// => [null]
 
_.castArray(undefined);
// => [undefined]
 
_.castArray();
// => []
```

`castArray` is the closest, except for null and undefined where we'd like an empty array

huafu commented on src/postprocess.ts:
```diff
@@ -52,70 +41,96 @@ function createBabelTransformer(
+  ];
+}
+
+function hasBabelAddon(inputList: any[], ...addonMatches: any[]): boolean {
```
IDK which node we target but I don't think node 6 do has `Array#includes`

ahnpnl replied:
we are safe to use `includes` with node 6. I just checked 
![image](https://user-images.githubusercontent.com/27772165/43354149-ed7e6c50-9246-11e8-9596-53050b1a4636.png)

GeeWee replied:
I was under the impression that they were always strings?

huafu commented on src/postprocess.ts:
```diff
@@ -52,70 +41,96 @@ function createBabelTransformer(
-    sourceMaps: tsJestConfig.disableSourceMapSupport !== true,
+    plugins: toArray(tsJestBabelConfig.plugins),
+    presets: toArray(tsJestBabelConfig.presets),
+    sourceMaps: tsJestConfig.disableSourceMapSupport ? false : 'both',
```
They are not emitted, but with a `true` instead of a `both`, it we won't return all the data jest wants. It's something that has changed in latest babel, but is backward compat since `both` is truthy.

GeeWee replied:
Yeah but emitted I don't mean written to disk - i mean added as inline and returned so we can pass it to jest.

huafu commented on src/postprocess.ts:
```diff
@@ -52,70 +41,96 @@ function createBabelTransformer(
-    codeSourcemapPair: CodeSourceMapPair,
+  // we need to know if there is istanbul plugin in the list so that we do not add it
+  // in the case the user already has it configured
+  const hasIstanbul = hasBabelAddon(plugins, istanbulPlugin);
```
well, I don't think many users have that of course, but I am doing myself in my `.babelrc.js` a `require` of some plugins and/or presets so that I can give myself some more options, and also use some old presets not compatible with new babel

GeeWee replied:
Alright fair enough

huafu commented:
- @GeeWee I think you'll definitely prefer [this](https://github.com/kulshekhar/ts-jest/pull/626/commits/fb5cd12a664b159ac841dfa3646f766a75f84134#diff-ae9757f32870dd5a0d6aafdc4d89af2d) as the test helper to mock jest config object

- also made the [test related to html-transform](https://github.com/huafu/ts-jest/blob/fb5cd12a664b159ac841dfa3646f766a75f84134/tests/__tests__/html-transform.spec.ts) much more readable (I hope at least)

- all the rest is just changes related to removing the ugly stuff about `__helpers__/jest-config.ts`

ahnpnl commented on src/utils/cache-file.ts:
```diff
@@ -0,0 +1,38 @@
+
+// FIXME: I think there is testRegexp and testMatch in Jest configugation
+// There is a default setting for that, but maybe it's defaulted from scratch
+export function cacheFile(
```
by default, `jest` provides default values for both [`testRegex` ](https://jestjs.io/docs/en/configuration#testregex-string)and [`testMatch`](https://jestjs.io/docs/en/configuration#testregex-string) and it doesn't have a default setting to say which one of them is the default setting to match test files because only one of them can be used at a time.

huafu replied:
Yup, I put that `FIXME` so that we can handle this in a future PR (implementation of that function hasn't changed). And I left `I think` to be sure the fixer will double check the input to fix it correctly and not ensure my comment was 100% right :-p

ahnpnl commented:
@huafu do we have a test for importing `json` files feature of typescript 2.9 ?

huafu commented:
@ahnpnl yup, simple, but it's [there](https://github.com/huafu/ts-jest/blob/fb5cd12a664b159ac841dfa3646f766a75f84134/tests/__tests__/import-json.spec.ts) and [there](https://github.com/huafu/ts-jest/tree/fb5cd12a664b159ac841dfa3646f766a75f84134/tests/import-json)

kulshekhar commented:
@GeeWee @ahnpnl I had taken a look at this PR last weekend and discussed it with @huafu 
I'm going to take some time out today to review this again but if you think this is in a good state, it's fine to merge it without waiting for me 

*In general, any one of us collaborators can merge PRs in if they're satisfied with it.*

kulshekhar commented on tests/__helpers__/test-utils.ts:
```diff
@@ -2,7 +2,7 @@ import * as fs from 'fs';
 import * as setFromArgv from 'jest-config/build/set_from_argv';
 import * as path from 'path';
-import { JestConfigNormalize } from '../../src/jest-types';
+import { JestConfigNormalize } from '../../dist/types';
```
this won't get the changed types right away, right?

huafu replied:
since they are test helpers, I must admit that it'd be more logic to take them from `src`, tho a build must be ran before we can start the tests anyway

kulshekhar replied:
scratch that - the project gets built before tests so this should be fine

kulshekhar commented on tests/__tests__/disable-sourcemap-support.spec.ts:
```diff
@@ -8,7 +8,8 @@ describe('Typescript errors', () => {
 
     // The actual error is on line 18 - the line# being wrong here means sourcemaps are disabled.
-    expect(stderr).toContain('Hello.ts:8');
+    expect(stderr).toMatch(/at .+Hello\.ts:\d+/);
```
This doesn't look right. Don't we want to check for exact line numbers to ensure that the sourcemaps are working correctly?

huafu replied:
it looks weird, I guess we should change the test description/comments. But here we are testing that, **without** sourcemaps, the line number reported is NOT the correct one, ie there is an error, but NOT on line 18. That is why I put 2 tests, first that there is an error at `Hello.ts` on some line, and then that the line number is NOT 18.

huafu replied:
https://github.com/huafu/ts-jest/blob/fb5cd12a664b159ac841dfa3646f766a75f84134/tests/__tests__/disable-sourcemap-support.spec.ts#L10

GeeWee commented:
Much nicer way of instantiating the mock config files - great job.
I had a few extra comments on a few pieces of code, but apart from that it looks fine.

Note that this will be a *breaking change* due to us forcing a beta version of babel. I have checked and babel-jest still uses the current version of babel (https://www.npmjs.com/package/babel-jest) - are we comfortable forcing users to switch to babel7?

ahnpnl commented:
My opinion is we should follow the babel version that jest is supporting.

huafu commented:
babel is not forced, they can install any version, well,  not 1.0.0 but you got me. The number i put in peerDeps is the recommended one.
i started another branch where the tests would run with different versions of babel and some other deps...

kulshekhar commented:
> babel is not forced, they can install any version

that was my understanding as well although tbh, I've never fully understood the finer points of deps in npm

ahnpnl commented:
I think putting `peerDependencies` like that is fine. It means that we are only compatible with that version in `peerDependencies`. Besides, we should add a note in `README` that if you are using babel, please make sure you have the same version as `peerDependencies`, otherwise you might encounter some unknown errors.
In general it is not a force, but it's a recommendation, but sometimes it can be understood as a force too. It fully depends on the purpose of the creator of the package and `README` guidance.

huafu commented:
@ahnpnl I totally agree with you, and I actually said to @kulshekhar that I'll create this troubleshooting page also talking about babel and other versions, linked in the readme (as I said in #635)... tho I did not do. I am starting it right now ;-)

huafu commented:
@kulshekhar @ahnpnl @GeeWee please review and update (not all at the same time for the update part haha): https://github.com/kulshekhar/ts-jest/wiki/Troubleshooting

huafu commented:
Hmmm weird, I did NOT create the `Home` wiki page, it sounds like GitHub auto-creates it if it does not exists and you create another page, I'll go update it as it sounds weird, just saw it now

huafu commented on README.md:
```diff
@@ -383,7 +377,6 @@ By default Jest ignores everything in `node_modules`. This setting prevents Jest
 ## Known Limitations
 ### Known limitations for TS compiler options
 - You can't use `"target": "ES6"` while using `node v4` in your test environment;
-- You can't use `"jsx": "preserve"` for now (see [progress of this issue](https://github.com/kulshekhar/ts-jest/issues/63));
```
The related issue is closed so I removed this

huafu commented on README.md:
```diff
@@ -1,17 +1,11 @@
-[![Build Status for linux](https://travis-ci.org/kulshekhar/ts-jest.svg?branch=master)](https://travis-ci.org/kulshekhar/ts-jest)
-[![Build Status for Windows](https://ci.appveyor.com/api/projects/status/g8tt9qd7usv0tolb/branch/master?svg=true)](https://ci.appveyor.com/project/kulshekhar/ts-jest/branch/master)
-
-**Important note: When using React 16 with Node version 8, you might see wrong line numbers for errors originating from tsx files. There's [an issue with more details on this](https://github.com/kulshekhar/ts-jest/issues/334)**
```
This has been moved to the troubleshooting page

huafu commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -0,0 +1,20 @@
+## Minimal repo <!-- If you haven't already, create the smallest possible repo that reproduces this issue by running `npm install` and `npm test`. This will speed up any fixes that this issue might need. Paste the minimal repo URL below -->:
+
+
+<!-- Optional (but highly recommended): Configure Travis (or your favorite system) with the minimal repo. This allows potential solutions to be tested against the minimal repo. This saves everyone time and avoids a lot of back and forth. -->
```
Here I've added the first comment. Also while keeping original texts I've put between comments the stuff that people use to leave in the issue text. Also changed from a bullet-list to bigger headers so that the issue is more readable.

GeeWee commented:
Did the babel npm package not update from 6 to 7? If there's no further code changes needed, I don't think there's any reason to update the peerDependency at all.

(Also let's not make this PR a moving target. It's awesome you want to keep contributing, but please make a new one)

huafu commented:
@GeeWee I am actually working on a new branch, here I added the links to the wiki page as I understood that it was kind of a missing thing for this PR to get merged. In the new branch I actually split the issue template in 2 so that people can select `issue`, `feature` or `other`, kind of as it's done [in jest repo](https://github.com/facebook/jest/issues/new/choose)

I did not get what you meant for the babel package. do you want me to revert the change related to it in `peerDependencies`?

GeeWee commented:
Ah, I see. That's awesome.

I meant, apart from the peerDependency warning, is this still going to work with babel 6?

huafu commented:
Oh @GeeWee I somehow missed your comment and was curious why I hadn't news.

I'll change that peer about babel, because it's wrong if we still allow babel 6. And I'll also rebase master. I've tested some mini repo which had babel-core v6 and it did work, but I'll also run another test to be sure.

huafu commented:
@GeeWee actually should I do a rebase or a simple merge? Asking since I think the rebase will vanish all the comments attached to diffs here right?

```
// git status after rebase:
On branch feature/upgrade-babel-and-fix-tsconfig
Your branch and 'origin/feature/upgrade-babel-and-fix-tsconfig' have diverged,
and have 29 and 22 different commits each, respectively.
```

kulshekhar commented:
A merge should be fine

On Thu 2 Aug, 2018, 11:05 Huafu Gandon, <notifications@github.com> wrote:

> @GeeWee <https://github.com/GeeWee> actually should I do a rebase or a
> simple merge? Asking since I think the rebase will vanish all the comments
> attached to diffs here right?
>
> // git status after rebase:
> On branch feature/upgrade-babel-and-fix-tsconfig
> Your branch and 'origin/feature/upgrade-babel-and-fix-tsconfig' have diverged,
> and have 29 and 22 different commits each, respectively.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kulshekhar/ts-jest/pull/626#issuecomment-409811252>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKR_0OmOxJGzOikYalQN68gzdiR1R6z9ks5uMo-4gaJpZM4VV7DG>
> .
>

huafu commented:
@kulshekhar can you double-check the overall changes in `README.md` and `.github/ISSUE_TEMPLATE.md` please?

ahnpnl commented on README.md:
```diff
@@ -404,7 +398,6 @@ By default Jest ignores everything in `node_modules`. This setting prevents Jest
 ## Known Limitations
 ### Known limitations for TS compiler options
 - You can't use `"target": "ES6"` while using `node v4` in your test environment;
```
I think we can put here that we don't support `node v4`.

huafu replied:
Agreed. Tho while updating the README I tried to keep my update as small as possible, and looked quickly to the stuff only related to adding the wiki troubleshooting link. Shouldn't we keep that for another PR?

GeeWee commented:
If @kulshekhar is good with the readme changes this looks good to me. Thanks for sticking with it.

ahnpnl commented:
103 comments 👍

kulshekhar commented:
lgtm as well

kulshekhar commented:
ts-jest doesn't follow semantic versioning. The only part that has rules is that the major version follows Jest's major version. The rest is arbitrary.

> Having both Babel 6 and Babel 7 as dependencies, with different tools using different versions, is definitely a bad decision.  The latest Jest also does not use @babel/core by default, so I'm curious where you're getting that information.

This PR addressed a lot of issues and so was accepted and merged. If you think it can be improved, a PR will be more than welcome.You're welcome to submit a patch.
